llmfoundry/utils/exceptions.py

```diff
@@ -28,46 +28,91 @@
     'InputFolderMissingDataError',
     'OutputFolderNotEmptyError',
     'MisconfiguredHfDatasetError',
+    'RunTimeoutError',
 ]
 
 ALLOWED_RESPONSE_KEYS = {'response', 'completion'}
 ALLOWED_PROMPT_KEYS = {'prompt'}
 ALLOWED_MESSAGES_KEYS = {'messages'}
 
-ErrorLocation = Union[Literal['TrainDataloader'], Literal['EvalDataloader']]
-ErrorAttribution = Union[Literal['UserError'], Literal['InternalError'],
-                         Literal['NetworkError']]
+FailureLocation = Union[Literal['TrainDataloader'], Literal['EvalDataloader']]
+FailureAttribution = Union[Literal['UserError'], Literal['InternalError'],
+                           Literal['NetworkError']]
 TrainDataLoaderLocation = 'TrainDataloader'
 EvalDataLoaderLocation = 'EvalDataloader'
 
 
-class ContextualError(Exception):
+class BaseContextualError(Exception):
     """Error thrown when an error occurs in the context of a specific task."""
 
-    location: Optional[ErrorLocation] = None
-    error_attribution: Optional[ErrorAttribution] = None
+    location: Optional[FailureLocation] = None
+    error_attribution: Optional[FailureAttribution] = None
 
+    def __init__(self, message: str, **kwargs: Any) -> None:
+        self.error = message
+        self.serializable_attributes = []
 
-class UserError(ContextualError):
+        for key, value in kwargs.items():
+            setattr(self, key, value)
+            self.serializable_attributes.append(key)
+
+        super().__init__(message)
+
+    def __reduce__(self):
```

Comment:
could you add a comment describing how we need this for pickling/multiprocessing from places like dataset.map